linera-witty-macros/src/wit_export/caller_type_parameter.rs

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Representation of a shared generic type parameter for the caller.
+
+use crate::util::{Specialization, Specializations};
+use proc_macro2::TokenStream;
+use proc_macro_error::abort;
+use quote::ToTokens;
+use std::collections::HashMap;
+use syn::{
+    punctuated::Punctuated, AngleBracketedGenericArguments, AssocType, GenericArgument, Generics,
+    Ident, PathArguments, PathSegment, PredicateType, Token, TraitBound, TraitBoundModifier, Type,
+    TypeParam, TypeParamBound, TypePath, WhereClause, WherePredicate,
+};
+
+/// Information on the  generic type parameter to use for the caller parameter, if present.
+#[derive(Clone, Copy, Debug)]
+pub enum CallerTypeParameter<'input> {
+    NotPresent,
+    WithoutUserData(&'input Ident),
+    WithUserData {
+        caller: &'input Ident,
+        user_data: &'input Type,
+    },
+}
+
+impl<'input> CallerTypeParameter<'input> {
+    /// Parses a type's [`Generics`] to determine if a caller type parameter should be used.
+    pub fn new(generics: &'input Generics) -> Self {
+        let where_bounds = Self::parse_bounds_from_where_clause(generics.where_clause.as_ref());
+
+        generics
+            .type_params()
+            .filter_map(|parameter| Self::try_from_parameter(parameter, &where_bounds))
+            .next()
+            .unwrap_or(CallerTypeParameter::NotPresent)
+    }
+
+    /// Parses the bounds present in an optional `where_clause`.
+    ///
+    /// Returns a map between constrained types and its predicates.
+    fn parse_bounds_from_where_clause(
+        where_clause: Option<&'input WhereClause>,
+    ) -> HashMap<&'input Ident, Vec<&'input TypeParamBound>> {
+        where_clause
+            .into_iter()
+            .flat_map(|where_clause| where_clause.predicates.iter())
+            .filter_map(|predicate| match predicate {
+                WherePredicate::Type(predicate) => Self::extract_predicate_bounds(predicate),
+                _ => None,
+            })
+            .collect()
+    }
```

Comment:
Is it okay for us to drop lifetime predicates?

Reply:
In this case it is, because we will only use the parsed bounds to search for a specific type constraint. The generic type parameter with that constraint (either implementing `Instance` trait or a trait that starts with `InstanceFor`) should be used as the caller type parameter.